# Data Preprocessing by Using Sklearn (scikit learn)

In [132]:
import seaborn as sns
import pandas as pd
from sklearn.preprocessing import StandardScaler, MinMaxScaler, LabelEncoder, OneHotEncoder
from sklearn.impute import SimpleImputer
from sklearn.model_selection import train_test_split
from sklearn.feature_selection import SelectKBest, chi2

In [133]:
# Load the dataset from the seaborn library 
titanic = sns.load_dataset("titanic")
titanic.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [134]:
titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


In [135]:
# Handle the missing values from the 'age' column
# Impute the missing values with the mean of the 'age' column by using SimpleImputer

# load the SimpleImputer from the from sklean.impute module of the sklearn library
from sklearn.impute import SimpleImputer

# initialize the imputer
imputer = SimpleImputer(strategy = 'mean')

# fit the imputer on the 'age' column
titanic['age'] = imputer.fit_transform(titanic[['age']])
titanic['age'].isnull().sum()

np.int64(0)

In [136]:
titanic.drop("deck" , axis= 1 , inplace=True)

In [137]:
titanic["fare"].max()

np.float64(512.3292)

In [138]:
titanic['fare'].min()

np.float64(0.0)

In [139]:
titanic["fare"].value_counts()

fare
8.0500     43
13.0000    42
7.8958     38
7.7500     34
26.0000    31
           ..
13.8583     1
50.4958     1
5.0000      1
9.8458      1
10.5167     1
Name: count, Length: 248, dtype: int64

In [140]:
shape_before = titanic.shape

# We see that the max value of the "fare" column is 512.3292 , which seems like a outlier 

so we use IQR (Inter Quartile Range) to calculate the outliers form this columns .

In [141]:
# Outlier Detection and Removal
# Detect and remove outliers in 'fare' based on the Interquartile Range (IQR)
Q1 = titanic['fare'].quantile(0.25)
Q3 = titanic['fare'].quantile(0.75)

IQR = Q3 - Q1
# calculate the lower and upper bounds for outliers
upper_bound = Q3 + IQR * 1.5
lower_bound = Q1 - IQR * 1.5

# remove outliers
titanic = titanic[(titanic['fare'] >= lower_bound) & (titanic['fare'] <= upper_bound)]

# check if any outliers were removed
shape_after = titanic.shape
total_outliers = shape_before[0] - shape_after[0]

print(f'Total number of outliers which were removed is :{total_outliers} ')

Total number of outliers which were removed is :116 


In [142]:
titanic['fare'].value_counts()

fare
8.0500     43
13.0000    42
7.8958     38
7.7500     34
26.0000    31
           ..
13.8583     1
50.4958     1
5.0000      1
9.8458      1
10.5167     1
Name: count, Length: 203, dtype: int64

In [143]:
titanic['fare'].max()

np.float64(65.0)

In [144]:
titanic['fare'].min()

np.float64(0.0)

In [145]:
# Normalization
# Normalize 'fare' to have values between 0 and 1
# Import the modul of MinMaxScaler from sklearn.preprocessing
from sklearn.preprocessing import MinMaxScaler

# Creating the object of MinMaxScaler or the initialization
min_max_scaler = MinMaxScaler()

# fit_transefer the MinMaxScaler on the 'fare' column
titanic['fare_normalized'] = min_max_scaler.fit_transform(titanic[['fare']])

titanic['fare'].value_counts()


fare
8.0500     43
13.0000    42
7.8958     38
7.7500     34
26.0000    31
           ..
13.8583     1
50.4958     1
5.0000      1
9.8458      1
10.5167     1
Name: count, Length: 203, dtype: int64

In [146]:
# Binning 
# Transform the 'age' column into three discrete categories

titanic['age_binned'] = pd.cut(titanic['age'], bins = [0, 18, 60, 100], labels= ['Child', 'Adult', 'Senior'])
titanic.sample(20) 

survived  pclass     sex        age  sibsp  parch     fare embarked  \
652         0       3    male  21.000000      0      0   8.4333        S   
593         0       3  female  29.699118      0      2   7.7500        Q   
636         0       3    male  32.000000      0      0   7.9250        S   
616         0       3    male  34.000000      1      1  14.4000        S   
217         0       2    male  42.000000      1      0  27.0000        S   
71          0       3  female  16.000000      5      2  46.9000        S   
343         0       2    male  25.000000      0      0  13.0000        S   
735         0       3    male  28.500000      0      0  16.1000        S   
751         1       3    male   6.000000      0      1  12.4750        S   
298         1       1    male  29.699118      0      0  30.5000        S   
381         1       3  female   1.000000      0      2  15.7417        C   
9           1       2  female  14.000000      1      0  30.0708        C   
855         1       3  female  18.000000      0      1   9.3500        S   
309         1       1  female  30.000000      0      0  56.9292        C   
41          0       2  female  27.000000      1      0  21.0000        S   
157         0       3    male  30.000000      0      0   8.0500        S   
249         0       2    male  54.000000      1      0  26.0000        S   
101         0       3    male  29.699118      0      0   7.8958        S   
286         1       3    male  30.000000      0      0   9.5000        S   
648         0       3    male  29.699118      0      0   7.5500        S   

      class    who  adult_male  embark_town alive  alone  fare_normalized  \
652   Third    man        True  Southampton    no   True         0.129743   
593   Third  woman       False   Queenstown    no  False         0.119231   
636   Third    man        True  Southampton    no   True         0.121923   
616   Third    man        True  Southampton    no  False         0.221538   
217  Second    man        True  Southampton    no  False         0.415385   
71    Third  woman       False  Southampton    no  False         0.721538   
343  Second    man        True  Southampton    no   True         0.200000   
735   Third    man        True  Southampton    no   True         0.247692   
751   Third  child       False  Southampton   yes  False         0.191923   
298   First    man        True  Southampton   yes   True         0.469231   
381   Third  child       False    Cherbourg   yes  False         0.242180   
9    Second  child       False    Cherbourg   yes  False         0.462628   
855   Third  woman       False  Southampton   yes  False         0.143846   
309   First  woman       False    Cherbourg   yes   True         0.875834   
41   Second  woman       False  Southampton    no  False         0.323077   
157   Third    man        True  Southampton    no   True         0.123846   
249  Second    man        True  Southampton    no  False         0.400000   
101   Third    man        True  Southampton    no   True         0.121474   
286   Third    man        True  Southampton   yes   True         0.146154   
648   Third    man        True  Southampton    no   True         0.116154   

    age_binned  
652      Adult  
593      Adult  
636      Adult  
616      Adult  
217      Adult  
71       Child  
343      Adult  
735      Adult  
751      Child  
298      Adult  
381      Child  
9        Child  
855      Child  
309      Adult  
41       Adult  
157      Adult  
249      Adult  
101      Adult  
286      Adult  
648      Adult

In [147]:
# Feature Engineering
# Create a new feature 'family_size' from 'sibsp' and 'parch'
titanic["family_size"] = titanic['sibsp'] + titanic['parch'] +1
titanic.head()

survived  pclass     sex        age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.000000      1      0   7.2500        S  Third   
2         1       3  female  26.000000      0      0   7.9250        S  Third   
3         1       1  female  35.000000      1      0  53.1000        S  First   
4         0       3    male  35.000000      0      0   8.0500        S  Third   
5         0       3    male  29.699118      0      0   8.4583        Q  Third   

     who  adult_male  embark_town alive  alone  fare_normalized age_binned  \
0    man        True  Southampton    no  False         0.111538      Adult   
2  woman       False  Southampton   yes   True         0.121923      Adult   
3  woman       False  Southampton   yes  False         0.816923      Adult   
4    man        True  Southampton    no   True         0.123846      Adult   
5    man        True   Queenstown    no   True         0.130128      Adult   

   family_size  
0            2  
2            1  
3            2  
4            1  
5            1

In [148]:
# Feature Selection
# Select the top 3 features that have the highest correlation with 'survived'
X = titanic[['pclass', 'age', 'sibsp', 'parch', 'fare_normalized']]
y = titanic['survived']
selector = SelectKBest(score_func=chi2, k=3)
X_selected = selector.fit_transform(X, y)